hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/TestDFSMkdirs.java

```diff
@@ -152,4 +154,80 @@ public void testMkdirRpcNonCanonicalPath() throws IOException {
       cluster.shutdown();
     }
   }
+
+  @Test
+  public void testMkDirsWithRestart() throws IOException {
+    MiniDFSCluster cluster =
+        new MiniDFSCluster.Builder(conf).numDataNodes(2).build();
+    DistributedFileSystem dfs = cluster.getFileSystem();
+    try {
+      // Create a dir in root dir, should succeed
+      assertTrue(dfs.mkdir(new Path("/mkdir-1"), FsPermission.getDefault()));
+      dfs.mkdir(new Path("/mkdir-2"), FsPermission.getDefault());
+      dfs.mkdir(new Path("/mkdir-3"), FsPermission.getDefault());
+      DFSTestUtil.writeFile(dfs, new Path("/mkdir-1/file1"), "hello world");
+      cluster.restartNameNodes();
+      dfs = cluster.getFileSystem();
+      assertTrue(dfs.exists(new Path("/mkdir-1")));
+      assertTrue(dfs.exists(new Path("/mkdir-2")));
+      assertTrue(dfs.exists(new Path("/mkdir-3")));
+      assertTrue(dfs.exists(new Path("/mkdir-1/file1")));
+    } finally {
+      dfs.close();
+      cluster.shutdown();
+    }
+  }
+
+  @Test
+  public void testMkdirWithDelete() throws IOException {
+    MiniDFSCluster cluster =
+        new MiniDFSCluster.Builder(conf).numDataNodes(2).build();
+    DistributedFileSystem dfs = cluster.getFileSystem();
+    // Create a dir in root dir, should succeed
+    String dirA = "/A";
+    String dirB = "/B";
+
+    String fileA = "/a";
+    String fileB = "/b";
+
+    try {
+      FsPermission fsP = FsPermission.getDefault();
+      dfs.mkdir(new Path(dirA), fsP);
+      dfs.mkdir(new Path(dirB), fsP);
+      dfs.mkdirs(new Path(dirB + "/B1/B2/B3"), fsP);
+
+      DFSTestUtil.writeFile(dfs, new Path(dirA + fileA), "hello world");
```

Comment:
This should be `new Path(dirA, fileA)`.
You don't want to deal with multi-OS file delimiters here.